README.md
# NAME

Task::EGA - EGA bundle

# SYNOPSIS

    cpanm --mirror-only --mirror https://stratopan.com/wangq/ega/master --interactive Task::EGA

# AUTHOR

Qiang Wang &lt;wang-q@outlook.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Qiang Wang.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Task::EGA

{{$NEXT}}

0.2.2 2017-01-13T16:04:11Z
    - More my modules

0.2.1 2016-12-16T00:42:59Z
    - Switch to Minilla

0.2.0  2016-12-12 00:37:43 CST
    - Make non-CPAN modules optional
    - Remove File::Sort, File::Remove, File::Rename, List::Flatten, DBIx::Class::Helpers, DBM::Deep,
      Statistics::Lite and Statistics::TTest

0.1.8  2016-12-10 00:42:52 CST
    - Remove MooseX::AttributeHelpers

0.1.7  2016-05-21 04:56:29 CST
    - Remove Bio::DB::EUtilities

0.1.6  2016-05-10 21:37:29 CST
    - Remove AlignDB::Util

0.1.5  2016-01-20 22:45:08 CST
    - Remove version requirement on some modules

0.1.4  2016-01-19 10:15:23 CST
    - Install all modules

0.1.3  2016-01-19 09:50:30 CST
    - A bunch of extra modules

0.1.2  2016-01-19 09:22:20 CST
    - Stratopan hasn't some new releases.

0.1.1  2016-01-19 08:57:19 CST
    - Stick to Bio::DB::EUtilities 1.72
    - Specify other modules' versions

0.0.1  2016-01-16 02:24:29 CST
    - Initial version
    - Modified from Task::Plack


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.006.
use strict;
use warnings;



use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "EGA bundle",
  "AUTHOR" => "Qiang Wang <wang-q\@outlook.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Task-EGA",
  "LICENSE" => "perl",
  "NAME" => "Task::EGA",
  "PREREQ_PM" => {
    "Algorithm::Munkres" => 0,
    "AlignDB::IntSpan" => 0,
    "AlignDB::SQL" => 0,
    "AlignDB::Stopwatch" => 0,
    "AnyEvent" => 0,
    "App::Ack" => 0,
    "App::Cmd" => "0.33",
    "App::Fasops" => 0,
    "App::RL" => 0,
    "App::Rangeops" => 0,
    "App::pmuninstall" => 0,
    "Archive::Extract" => 0,
    "Array::Compare" => 0,
    "Bio::ASN1::EntrezGene" => 0,
    "Bio::Graphics" => 0,
    "Bio::Phylo" => 0,
    "Bio::Seq" => 0,
    "Bio::Tools::Run::Alignment::Clustalw" => 0,
    "CPAN" => "1.81",
    "CPAN::Mini" => 0,
    "CPANDB" => 0,
    "Cairo" => 0,
    "Cairo::GObject" => 0,
    "Chart::Math::Axis" => 0,
    "Clone" => 0,
    "Config::General" => 0,
    "Config::Tiny" => 0,
    "Convert::Binary::C" => 0,
    "DBD::CSV" => 0,
    "DBD::mysql" => 0,
    "DBI" => 0,
    "DB_File" => 0,
    "Data::Dumper" => 0,
    "Data::Stag" => "0.11",
    "DateTime::Format::Natural" => 0,
    "Devel::CheckLib" => 0,
    "Dist::Milla" => 0,
    "EV" => 0,
    "Error" => 0,
    "Excel::Writer::XLSX" => "0.88",
    "ExtUtils::Manifest" => "1.52",
    "File::Copy::Recursive" => 0,
    "File::Find::Rule" => 0,
    "File::HomeDir" => 0,
    "File::Listing" => 0,
    "Font::TTF::Font" => 0,
    "GD" => 0,
    "GD::SVG" => 0,
    "Getopt::Long::Descriptive" => 0,
    "Glib" => 0,
    "Glib::Object::Introspection" => 0,
    "Graph" => 0,
    "Gtk3" => 0,
    "HTML::TableExtract" => 0,
    "IO::All" => 0,
    "IO::All::LWP" => 0,
    "IO::String" => 0,
    "Import::Into" => 0,
    "JSON::XS" => 0,
    "LWP::Online" => 0,
    "LWP::Protocol::https" => 0,
    "LWP::Protocol::socks" => 0,
    "List::MoreUtils" => 0,
    "MCE" => "1.81",
    "Math::Bezier" => 0,
    "Math::BigFloat" => 0,
    "Math::BigInt" => 0,
    "Math::CDF" => 0,
    "Math::GSL" => 0,
    "Math::Random" => 0,
    "Math::Random::MT::Auto" => 0,
    "Math::Round" => 0,
    "Math::VecStat" => 0,
    "Module::Build" => "0.42",
    "Module::Path" => 0,
    "Mojolicious" => 0,
    "MongoDB" => 0,
    "Moo" => 0,
    "MooX::Options" => 0,
    "Moose" => 0,
    "MooseX::Storage" => 0,
    "PDF::API2" => 0,
    "PDL" => 0,
    "POE" => 0,
    "PPI::XS" => 0,
    "Parallel::ForkManager" => 0,
    "Params::Validate" => 0,
    "Path::Class" => 0,
    "Path::Tiny" => 0,
    "Perl::Tidy" => 0,
    "Pod::POM::Web" => 0,
    "PostScript::TextBlock" => 0,
    "Proc::Background" => 0,
    "Readonly" => 0,
    "Regexp::Common" => 0,
    "Roman" => 0,
    "SVG" => 0,
    "SVG::Graph" => 0,
    "Scalar::Util" => 0,
    "Search::Indexer" => 0,
    "Set::IntSpan" => 0,
    "Set::Scalar" => 0,
    "Sort::Naturally" => 0,
    "Spreadsheet::ParseExcel" => 0,
    "Spreadsheet::WriteExcel" => 0,
    "Statistics::Basic" => 0,
    "Statistics::R" => 0,
    "String::Compare" => 0,
    "Template" => 0,
    "Test::Class" => 0,
    "Test::Harness" => "2.62",
    "Test::Most" => 0,
    "Test::Roo" => 0,
    "Test::Taint" => 0,
    "Test::Without::Module" => 0,
    "Text::Balanced" => 0,
    "Text::CSV_XS" => 0,
    "Text::Format" => 0,
    "Text::Table" => 0,
    "Tie::IxHash" => 0,
    "Time::Duration" => 0,
    "URI::Escape" => 0,
    "WWW::Mechanize" => 0,
    "XML::DOM::XPath" => 0,
    "XML::LibXML" => 0,
    "XML::SAX::Writer" => 0,
    "XML::Simple" => 0,
    "XML::Twig" => 0,
    "XML::Writer" => 0,
    "XML::XML2JSON" => 0,
    "YAML" => 0,
    "YAML::Syck" => 0
  },
  "VERSION" => "0.2.0",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Algorithm::Munkres" => 0,
  "AlignDB::IntSpan" => 0,
  "AlignDB::SQL" => 0,
  "AlignDB::Stopwatch" => 0,
  "AnyEvent" => 0,
  "App::Ack" => 0,
  "App::Cmd" => "0.33",
  "App::Fasops" => 0,
  "App::RL" => 0,
  "App::Rangeops" => 0,
  "App::pmuninstall" => 0,
  "Archive::Extract" => 0,
  "Array::Compare" => 0,
  "Bio::ASN1::EntrezGene" => 0,
  "Bio::Graphics" => 0,
  "Bio::Phylo" => 0,
  "Bio::Seq" => 0,
  "Bio::Tools::Run::Alignment::Clustalw" => 0,
  "CPAN" => "1.81",
  "CPAN::Mini" => 0,
  "CPANDB" => 0,
  "Cairo" => 0,
  "Cairo::GObject" => 0,
  "Chart::Math::Axis" => 0,
  "Clone" => 0,
  "Config::General" => 0,
  "Config::Tiny" => 0,
  "Convert::Binary::C" => 0,
  "DBD::CSV" => 0,
  "DBD::mysql" => 0,
  "DBI" => 0,
  "DB_File" => 0,
  "Data::Dumper" => 0,
  "Data::Stag" => "0.11",
  "DateTime::Format::Natural" => 0,
  "Devel::CheckLib" => 0,
  "Dist::Milla" => 0,
  "EV" => 0,
  "Error" => 0,
  "Excel::Writer::XLSX" => "0.88",
  "ExtUtils::Manifest" => "1.52",
  "File::Copy::Recursive" => 0,
  "File::Find::Rule" => 0,
  "File::HomeDir" => 0,
  "File::Listing" => 0,
  "Font::TTF::Font" => 0,
  "GD" => 0,
  "GD::SVG" => 0,
  "Getopt::Long::Descriptive" => 0,
  "Glib" => 0,
  "Glib::Object::Introspection" => 0,
  "Graph" => 0,
  "Gtk3" => 0,
  "HTML::TableExtract" => 0,
  "IO::All" => 0,
  "IO::All::LWP" => 0,
  "IO::String" => 0,
  "Import::Into" => 0,
  "JSON::XS" => 0,
  "LWP::Online" => 0,
  "LWP::Protocol::https" => 0,
  "LWP::Protocol::socks" => 0,
  "List::MoreUtils" => 0,
  "MCE" => "1.81",
  "Math::Bezier" => 0,
  "Math::BigFloat" => 0,
  "Math::BigInt" => 0,
  "Math::CDF" => 0,
  "Math::GSL" => 0,
  "Math::Random" => 0,
  "Math::Random::MT::Auto" => 0,
  "Math::Round" => 0,
  "Math::VecStat" => 0,
  "Module::Build" => "0.42",
  "Module::Path" => 0,
  "Mojolicious" => 0,
  "MongoDB" => 0,
  "Moo" => 0,
  "MooX::Options" => 0,
  "Moose" => 0,
  "MooseX::Storage" => 0,
  "PDF::API2" => 0,
  "PDL" => 0,
  "POE" => 0,
  "PPI::XS" => 0,
  "Parallel::ForkManager" => 0,
  "Params::Validate" => 0,
  "Path::Class" => 0,
  "Path::Tiny" => 0,
  "Perl::Tidy" => 0,
  "Pod::POM::Web" => 0,
  "PostScript::TextBlock" => 0,
  "Proc::Background" => 0,
  "Readonly" => 0,
  "Regexp::Common" => 0,
  "Roman" => 0,
  "SVG" => 0,
  "SVG::Graph" => 0,
  "Scalar::Util" => 0,
  "Search::Indexer" => 0,
  "Set::IntSpan" => 0,
  "Set::Scalar" => 0,
  "Sort::Naturally" => 0,
  "Spreadsheet::ParseExcel" => 0,
  "Spreadsheet::WriteExcel" => 0,
  "Statistics::Basic" => 0,
  "Statistics::R" => 0,
  "String::Compare" => 0,
  "Template" => 0,
  "Test::Class" => 0,
  "Test::Harness" => "2.62",
  "Test::Most" => 0,
  "Test::Roo" => 0,
  "Test::Taint" => 0,
  "Test::Without::Module" => 0,
  "Text::Balanced" => 0,
  "Text::CSV_XS" => 0,
  "Text::Format" => 0,
  "Text::Table" => 0,
  "Tie::IxHash" => 0,
  "Time::Duration" => 0,
  "URI::Escape" => 0,
  "WWW::Mechanize" => 0,
  "XML::DOM::XPath" => 0,
  "XML::LibXML" => 0,
  "XML::SAX::Writer" => 0,
  "XML::Simple" => 0,
  "XML::Twig" => 0,
  "XML::Writer" => 0,
  "XML::XML2JSON" => 0,
  "YAML" => 0,
  "YAML::Syck" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/Task/EGA.pm
package Task::EGA;
use strict;
use warnings;
use autodie;

our $VERSION = '0.2.2';

sub dependencies {
    return (
        # EGA
        [   'Basic set of tools',
            1,
            [   ['Archive::Extract'],          ['Config::Tiny'],
                ['DB_File'],                   ['File::Find::Rule'],
                ['Getopt::Long::Descriptive'], ['List::MoreUtils'],
                ['JSON::XS'],                  ['Path::Tiny'],
                ['Text::CSV_XS'],              ['Tie::IxHash'],
                ['YAML'],                      ['YAML::Syck'],
            ],
        ],
        [   'Basic set of tools with huge tests',
            1,
            [   [ 'AnyEvent',       0 ],
                [ 'App::Ack',       0 ],
                [ 'App::Cmd',       0.330 ],
                [ 'DBI',            0 ],
                [ 'EV',             0 ],
                [ 'MCE',            1.810 ],
                [ 'Moo',            0 ],
                [ 'Moose',          0 ],
                [ 'Perl::Tidy',     0 ],
                [ 'POE',            0 ],
                [ 'Template',       0 ],
                [ 'WWW::Mechanize', 0 ],
            ],
        ],
        [ 'GD', 1, [ [ 'GD', 0 ], ['SVG'], ['GD::SVG'], ], ],
        [   'BioPerl needed',
            1,
            [
                # requires
                [ 'IO::String',         0 ],
                [ 'Data::Stag',         0.11 ],
                [ 'Scalar::Util',       0 ],
                [ 'ExtUtils::Manifest', 1.52 ],

                # configure_requires
                [ 'Module::Build', 0.42 ],

                # build_requires
                [ 'CPAN',          1.81 ],
                [ 'Test::Harness', 2.62 ],
                [ 'Test::Most',    0 ],
                [ 'URI::Escape',   0 ],
            ],
        ],
        [   'BioPerl recommended',
            1,
            [   ['Algorithm::Munkres'], ['Array::Compare'],
                ['Clone'],              ['Convert::Binary::C'],
                ['Error'],              ['Graph'],
                ['HTML::TableExtract'], ['PostScript::TextBlock'],
                ['Set::Scalar'],        ['Sort::Naturally'],
                ['SVG::Graph'],         ['XML::DOM::XPath'],
                ['XML::LibXML'],        ['XML::SAX::Writer'],
                ['XML::Simple'],        ['XML::Twig'],
                ['XML::Writer'],
            ],
        ],
        [   'BioPerl',
            1,
            [ [ 'Bio::Seq', 0 ], [ 'Bio::ASN1::EntrezGene', 0 ], ],
        ],
        [   'Bio::Phylo',
            1,
            [   ['Math::CDF'], ['Math::Random'], ['PDF::API2'], ['XML::XML2JSON'],
                [ 'Bio::Phylo', 0 ],
            ],
        ],
        [   'circos',
            1,
            [   ['Config::General'], ['Data::Dumper'],
                ['Math::Bezier'],    ['Math::BigInt'],
                ['Math::BigFloat'],  ['Math::Round'],
                ['Math::VecStat'],   ['Params::Validate'],
                ['Readonly'],        ['Regexp::Common'],
                ['Set::IntSpan'],    ['Statistics::Basic'],
                ['Text::Balanced'],  ['Text::Format'],
                [ 'Font::TTF::Font', 0 ],
            ],
        ],
        [   'Others',
            1,
            [   ['DateTime::Format::Natural'], ['File::HomeDir'],
                ['DBD::CSV'],                  ['Path::Class'],
                ['File::Listing'],             ['File::Copy::Recursive'],
                [ 'String::Compare',      0 ], [ 'LWP::Online',          0 ],
                [ 'LWP::Protocol::https', 0 ], [ 'LWP::Protocol::socks', 0 ],
            ],
        ],
        [   'My own modules',
            1,
            [   [ 'AlignDB::IntSpan',   0 ],
                [ 'AlignDB::Stopwatch', 0 ],
                [ 'AlignDB::SQL',       0 ],
                [ 'AlignDB::Codon',     0 ],
                [ 'AlignDB::Window',    0 ],
                [ 'AlignDB::DeltaG',    0 ],
                [ 'AlignDB::GC',        0 ],
                [ 'AlignDB::ToXLSX',    0 ],
                [ 'AlignDB::Run',       0 ],
                [ 'App::Fasops',        0 ],
                [ 'App::RL',            0 ],
                [ 'App::Rangeops',      0 ],
            ],
        ],

        # AlignDB
        [   'Bio::Graphics and BioPerl-run',
            1, [ [ 'Bio::Graphics', 0 ], [ 'Bio::Tools::Run::Alignment::Clustalw', 0 ], ],
        ],
        [   'Test::*', 1,
            [ ['Test::Class'], ['Test::Roo'], ['Test::Taint'], ['Test::Without::Module'], ],
        ],
        [ 'Moose related', 1, [ ['MooX::Options'], ['MooseX::Storage'], ], ],
        [   'Develop',
            1,
            [   ['App::pmuninstall'], ['CPANDB'],
                ['Minilla'],          ['Module::Path'],
                ['Pod::POM::Web'],    ['Search::Indexer'],
                ['Proc::Background'], ['Devel::CheckLib'],
                ['Import::Into'],     ['PPI::XS'],
            ],
        ],
        [   'Gtk3 stuffs',
            1,
            [   [ 'Glib',                        0 ],
                [ 'Cairo',                       0 ],
                [ 'Cairo::GObject',              0 ],
                [ 'Glib::Object::Introspection', 0 ],
                [ 'Gtk3',                        0 ],
            ],
        ],
        [   'Database and WWW',
            1,
            [   [ 'DBD::mysql',   0 ],
                [ 'MongoDB',      0 ],
                [ 'Mojolicious',  0 ],
                [ 'IO::All',      0 ],
                [ 'IO::All::LWP', 0 ],
            ],
        ],
        [   'text and xlsx',
            1,
            [   ['Roman'],             ['Text::Table'],
                ['Time::Duration'],    [ 'Excel::Writer::XLSX', 0.88 ],
                ['Spreadsheet::XLSX'], ['Chart::Math::Axis'],
            ],
        ],
        [   'Math and Stat',
            1,
            [   ['Math::Random::MT::Auto'],
                [ 'PDL',           0 ],
                [ 'Math::GSL',     0 ],
                [ 'Statistics::R', 0 ],
            ],
        ],

        # [ 'Extra set of my own modules', 0, [], ],
    );
}

sub iter_deps {
    my ( $class, $cb ) = @_;
    my @deps = $class->dependencies;
    while ( my $ref = splice @deps, 0, 1 ) {
        $cb->( @{$ref} );
    }
}

sub cpanfile {
    my $class   = shift;
    my $fh      = shift;
    my $current = shift;

    $class->iter_deps(
        sub {
            my ( $name, $cond, $deps ) = @_;

            if ($cond) {
                $fh->print("# $name\n");
            }
            else {
                ( my $ident = $name ) =~ s/[^A-Za-z_]+/_/g;
                $fh->print( "feature '", lc($ident), "', '$name' => sub {\n" );
            }

            for my $dep ( @{$deps} ) {
                my ( $module, $version ) = @{$dep};
                if ( !defined($version) and $current ) {
                    $version = version_for($module);
                }

                if ( !$cond ) {
                    $fh->print("    ");
                }

                if ($version) {
                    $fh->printf( "requires '%s', '%s';\n", $module, $version );
                }
                else {
                    $fh->printf( "requires '%s';\n", $module );
                }
            }

            if ( !$cond ) {
                $fh->print("};\n");
            }
            $fh->print("\n");
        }
    );
}

sub version_for {
    my $dist = shift;

    ( my $module = $dist ) =~ s/-/::/g;
    my $info = `cpanm --info $module ` or return;
    return ( $info =~ /-([\d\.]+)\.tar\.gz/ )[0];
}

1;

__END__

=encoding utf-8

=for stopwords

=head1 NAME

Task::EGA - EGA bundle

=head1 SYNOPSIS

    cpanm --mirror-only --mirror https://stratopan.com/wangq/ega/master --interactive Task::EGA

=head1 AUTHOR

Qiang Wang <wang-q@outlook.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Qiang Wang.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
